x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,84 +358,172 @@ public void removeFromCache(final CacheKey cacheKey) {
-     * @param shardId    the searchable snapshots shard id
-     * @param runnable   a runnable to execute
+     * @param snapshotUUID      the snapshot's unique identifier
+     * @param snapshotIndexName the name of the index in the snapshot
+     * @param shardId           the {@link ShardId}
      */
-    public void runIfShardMarkedAsEvictedInCache(SnapshotId snapshotId, IndexId indexId, ShardId shardId, Runnable runnable) {
-        runIfShardMarkedAsEvictedInCache(new ShardEviction(snapshotId.getUUID(), indexId.getName(), shardId), runnable);
+    public void evictCacheFilesIfNeeded(String snapshotUUID, String snapshotIndexName, ShardId shardId) {
+        final ShardEviction shardEviction = new ShardEviction(snapshotUUID, snapshotIndexName, shardId);
+        FutureUtils.get(processShardEvictionIfNeeded(shardEviction));
     }
 
     /**
-     * Allows to run the specified {@link Runnable} if the shard represented by {@link ShardEviction} is still marked as evicted at the time
-     * this method is executed. The @link Runnable} will be executed while the current thread is holding the lock associated to the shard.
+     * Evicts the cache files associated to the specified {@link ShardEviction} if the shard is marked as evicted in cache at the time
+     * this method is executed. Multiple threads can attempt to process the same shard eviction at the same time but only one thread will
+     * evict the cache files and other threads can use the returned {@link CompletableFuture} to check if the computation is complete
+     * or to wait for its completion.
      *
-     * @param shardEviction a {@link ShardEviction} representing the shard marked as evicted
-     * @param runnable      a runnable to execute
+     * @param shardEviction the shard eviction to process
+     * @return a {@link CompletableFuture} that can be used to wait for the eviction to complete
      */
-    private void runIfShardMarkedAsEvictedInCache(ShardEviction shardEviction, Runnable runnable) {
-        try (Releasable ignored = shardsEvictionLock.acquire(shardEviction)) {
-            boolean success = false;
-            try {
-                if (evictedShards.remove(shardEviction)) {
-                    runnable.run();
+    CompletableFuture<Integer> processShardEvictionIfNeeded(ShardEviction shardEviction) {
+        final CompletableFuture<Integer> completableFuture;
+        synchronized (shardsEvictionsMutex) {
+            assert assertGenericThreadPool();
+            assert shardsEvictionsInvariant();
+
+            if (pendingShardsEvictions.contains(shardEviction) == false) {
+                logger.trace("shard [{}] not marked as evicted", shardEviction);
+                assert runningShardsEvictions.get(shardEviction) == null : "found a running shard eviction for " + shardEviction;
+                return CompletableFuture.completedFuture(0);
+
+            } else if (runningShardsEvictions.containsKey(shardEviction)) {
```

Comment:
I think this could just be simplified with `computeIfAbsent` in an `else` branch?